src/Stratis.Bitcoin.Features.Consensus/PosCoinviewRule.cs

```diff
@@ -45,92 +20,77 @@ public class PosConsensusValidator : PowConsensusValidator, IPosConsensusValidat
         /// <summary>Drifting Bug Fix, hardfork on Sat, 19 Nov 2016 00:00:00 GMT.</summary>
         public const long DriftingBugFixTimestamp = 1479513600;
 
-        /// <summary>Instance logger.</summary>
-        private readonly ILogger logger;
-
         /// <summary>Provides functionality for checking validity of PoS blocks.</summary>
-        public IStakeValidator StakeValidator { get; }
+        public IStakeValidator StakeValidator { get; set; }
 
         /// <summary>Database of stake related data for the current blockchain.</summary>
-        private readonly IStakeChain stakeChain;
+        private IStakeChain stakeChain;
+
+        private PosConsensusOptions posConsensusOptions;
+
+        public override void Initialize()
+        {
+            base.Initialize();
 
-        /// <summary>Proof of Stake consensus options.</summary>
-        private readonly PosConsensusOptions consensusOptions;
+            var consensusRules = (PosConsensusRules)this.Parent;
+
+            this.StakeValidator = consensusRules.StakeValidator;
+            this.stakeChain = consensusRules.StakeChain;
+            this.posConsensusOptions = (PosConsensusOptions)this.ConsensusOptions;
+        }
 
         /// <inheritdoc />
-        /// <param name="stakeValidator">Provides functionality for checking validity of PoS blocks.</param>
-        /// <param name="checkpoints">Provider of block header hash checkpoints.</param>
-        /// <param name="network">Specification of the network the node runs on - regtest/testnet/mainnet.</param>
-        /// <param name="stakeChain">Database of stake related data for the current blockchain.</param>
-        /// <param name="dateTimeProvider">Provider of time functions.</param>
-        /// <param name="loggerFactory">Factory for creating loggers.</param>
-        public PosConsensusValidator(
-            IStakeValidator stakeValidator,
-            ICheckpoints checkpoints,
-            Network network,
-            IStakeChain stakeChain,
-            IDateTimeProvider dateTimeProvider,
-            ILoggerFactory loggerFactory)
-            : base(network, checkpoints, dateTimeProvider, loggerFactory)
+        public override Task RunAsync(RuleContext context)
         {
-            Guard.NotNull(network.Consensus.Option<PosConsensusOptions>(), nameof(network.Consensus.Options));
+            this.Logger.LogTrace("()");
+
+            // Compute and store the stake proofs.
+            this.CheckAndComputeStake(context);
 
-            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
-            this.StakeValidator = stakeValidator;
-            this.stakeChain = stakeChain;
-            this.consensusOptions = network.Consensus.Option<PosConsensusOptions>();
+            base.RunAsync(context);
+
+            this.stakeChain.Set(context.BlockValidationContext.ChainedHeader, context.Stake.BlockStake);
```

Comment:
.Set here is 

```
public void Set(ChainedHeader chainedHeader, BlockStake blockStake)
{
    this.SetAsync(chainedHeader, blockStake).GetAwaiter().GetResult();
}
```

So since we are calling it from an async method maybe it will be a good idea to remove .Set at all and use SetAsync instead?

Reply:
@majikandy please implement this as well.

Reply:
Pretty much, yes

Reply:
cool, thanks - that will be in the next commit

Reply:
I'm also changing the PowCoinviewRule  part where it was:-

`bool passed = checkInputs.All(c => c.GetAwaiter().GetResult());`

to

`bool passed = (await Task.WhenAll(checkInputs)).All(b => b);`

Then the PosCoinviewRule call down to base can be awaited too.